### PR TITLE
feat: add frequent typo-catching rules

### PR DESCRIPTION
**Pull request by jay7x**
As Salt is unable to catch some frequent typos there are few checks:
- "onchange" (must be "onchanges")
- "requires" (must be "require")
- "content" (must be "contents")

### CODE REVIEW

**Comment by jay7x** [comment]
I put those rules as 'formatting' though we may have 'typos' category maybe..

**Comment by jbouter** [comment]
:wave: Hi @jay7x! Thanks for your PR! We are currently a bit short on time, but we'll try to look at your code soon.

File: tests/unit/TestTypoOnchangesRule.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+# Copyright (c) 2021 Yury Bushmelev
+
+import unittest
+
+from saltlint.linter.collection import RulesCollection
+from saltlint.rules.TypoOnchangesRule import TypoOnchangesRule
+from tests import RunFromText
+
+
+GOOD_MODE_QUOTATION_LINE = '''
```
Comment (jbouter) [inline tests/unit/TestTypoOnchangesRule.py:14]:
```suggestion
GOOD_ONCHANGES_LINE = '''
```

File: tests/unit/TestTypoOnchangesRule.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    - onchanges:
+        - otherfile
+    - onchanges_in:
+        - anotherfile
+    - onchanges_any:
+        - yetanotherfile
+        - onemorefile
+'''
+
+BAD_MODE_QUOTATION_LINE = '''
```
Comment (jbouter) [inline tests/unit/TestTypoOnchangesRule.py:30]:
```suggestion
BAD_ONCHANGES_LINE = '''
```

File: tests/unit/TestTypoOnchangesRule.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+class TestTypoOnchangesRule(unittest.TestCase):
+    collection = RulesCollection()
+
+    def setUp(self):
+        self.collection.register(TypoOnchangesRule())
+        self.runner = RunFromText(self.collection)
+
+    def test_statement_positive(self):
+        results = self.runner.run_state(GOOD_MODE_QUOTATION_LINE)
```
Comment (jbouter) [inline tests/unit/TestTypoOnchangesRule.py:54]:
```suggestion
        results = self.runner.run_state(GOOD_ONCHANGES_LINE)
```

File: tests/unit/TestTypoOnchangesRule.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        self.assertEqual(0, len(results))
+
+    def test_statement_negative(self):
+        results = self.runner.run_state(BAD_MODE_QUOTATION_LINE)
```
Comment (jbouter) [inline tests/unit/TestTypoOnchangesRule.py:58]:
```suggestion
        results = self.runner.run_state(BAD_ONCHANGES_LINE)
```

File: tests/unit/TestTypoRequireRule.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+# Copyright (c) 2021 Yury Bushmelev
+
+import unittest
+
+from saltlint.linter.collection import RulesCollection
+from saltlint.rules.TypoRequireRule import TypoRequireRule
+from tests import RunFromText
+
+
+GOOD_MODE_QUOTATION_LINE = '''
```
Comment (jbouter) [inline tests/unit/TestTypoRequireRule.py:14]:
```suggestion
GOOD_REQUIRE_LINE = '''
```

File: tests/unit/TestTypoRequireRule.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    - require:
+        - otherfile
+    - require_in:
+        - anotherfile
+    - require_any:
+        - yetanotherfile
+        - onemorefile
+'''
+
+BAD_MODE_QUOTATION_LINE = '''
```
Comment (jbouter) [inline tests/unit/TestTypoRequireRule.py:30]:
```suggestion
BAD_REQUIRE_LINE = '''
```

File: tests/unit/TestTypoRequireRule.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+class TestTypoRequireRule(unittest.TestCase):
+    collection = RulesCollection()
+
+    def setUp(self):
+        self.collection.register(TypoRequireRule())
+        self.runner = RunFromText(self.collection)
+
+    def test_statement_positive(self):
+        results = self.runner.run_state(GOOD_MODE_QUOTATION_LINE)
```
Comment (jbouter) [inline tests/unit/TestTypoRequireRule.py:54]:
```suggestion
        results = self.runner.run_state(GOOD_REQUIRE_LINE)
```

File: tests/unit/TestTypoRequireRule.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+        self.assertEqual(0, len(results))
+
+    def test_statement_negative(self):
+        results = self.runner.run_state(BAD_MODE_QUOTATION_LINE)
```
Comment (jbouter) [inline tests/unit/TestTypoRequireRule.py:58]:
```suggestion
        results = self.runner.run_state(BAD_REQUIRE_LINE)
```

File: tests/unit/TestTypoContentsRule.py [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+# Copyright (c) 2021 Yury Bushmelev
+
+import unittest
+
+from saltlint.linter.collection import RulesCollection
+from saltlint.rules.TypoContentsRule import TypoContentsRule
+from tests import RunFromText
+
+
+GOOD_MODE_QUOTATION_LINE = '''
```
Comment (jbouter) [inline tests/unit/TestTypoContentsRule.py:14]:
```suggestion
GOOD_CONTENTS_LINE = '''
```

File: tests/unit/TestTypoContentsRule.py [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    - group: root
+    - mode: '0700'
+    - contents: test
+    - contents_pillar: foo
+    - contents_grains: bar
+    - contents_newline: True
+    - contents_delimiter: '!'
+'''
+
+BAD_MODE_QUOTATION_LINE = '''
```
Comment (jbouter) [inline tests/unit/TestTypoContentsRule.py:28]:
```suggestion
BAD_CONTENTS_LINE = '''
```

File: tests/unit/TestTypoContentsRule.py [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+
+class TestTypoContentsRule(unittest.TestCase):
+    collection = RulesCollection()
+
+    def setUp(self):
+        self.collection.register(TypoContentsRule())
+        self.runner = RunFromText(self.collection)
+
+    def test_statement_positive(self):
+        results = self.runner.run_state(GOOD_MODE_QUOTATION_LINE)
```
Comment (jbouter) [inline tests/unit/TestTypoContentsRule.py:50]:
```suggestion
        results = self.runner.run_state(GOOD_CONTENTS_LINE)
```

File: tests/unit/TestTypoContentsRule.py [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+        self.assertEqual(0, len(results))
+
+    def test_statement_negative(self):
+        results = self.runner.run_state(BAD_MODE_QUOTATION_LINE)
```
Comment (jbouter) [inline tests/unit/TestTypoContentsRule.py:54]:
```suggestion
        results = self.runner.run_state(BAD_CONTENTS_LINE)
```

**Review by jbouter — COMMENTED** [review]
Some minor suggestions :-)

**Comment by jay7x** [comment]
Ah! I see.. I missed that BAD_XXX_LINE & GOOD_XXX_LINE are different for every test somehow :-D

**Review by jbouter — APPROVED** [review]

**Comment by jbouter** [comment]
Approved on my end, but I'd like a review by @roaldnefs as well.

**Comment by jay7x** [comment]
Any chance for this PR to be merged or declined soon? :-D

**Review by roaldnefs — APPROVED** [review]
Besides the missing `CHANGELOG.md` entry I've made some minor changes by adding a `TypographicalErrorRule` base rule that can easily be extended by any of the typographical error rules.

**Pull request merged**